### DISCUSSION TITLE
GHA tests that depend on Postgres fail intermittently:

### DISCUSSION THREAD

**Discussion by mattsimgithub** [community/community#25757] [Actions]
<p>Our GH workflow is configured to run tests automatically upon pushes.</p>
<p>Recently, I switched our test backend db on Github Actions from SQLite to Postgres. Since then we get errors like this, often but not every time:</p>
<pre><code class="lang-auto">psycopg2.OperationalError: could not connect to server: Connection refused
	Is the server running on host "db" (172.19.0.2) and accepting
	TCP/IP connections on port 5432?
</code></pre>
<p>Sometimes the jobs work the first time, or I can re-run the same failing job some hours later and it will succeed. The failures are intermittent.</p>
<p>The YAML for the GH workflow is here: <a href="https://pastebin.com/Xhy1sssG" class="inline-onebox" rel="noopener nofollow ugc">failng-pg-jobs - Pastebin.com</a></p>
<p>I find a discussion here <a href="https://stackoverflow.com/questions/55190058/django-was-unable-to-create-a-connection-to-the-postgres-database-and-will-use" class="inline-onebox" rel="noopener nofollow ugc">python 3.x - Django was unable to create a connection to the 'postgres' database and will use the default database instead - Stack Overflow</a> where someone had a similar issue and was advised to introduce a package called <code>wait-for-it</code> to ensure the db was fully initialized.</p>
<p>I can try this, and will if I don’t find any bettter ideas, but I’m using <code>--health-cmd pg_isready</code> already (see the YAML above) and I’m not crazy about just throwing more tools for delay into the mix.</p>
<p>Anybody have ideas or suggestions?</p>

**Comment by mattsimgithub** [comment, marked as answer]
<p>This seems to be resolved (knock on wood).</p>
<p>What ended up working was using <strong>depends_on</strong> together with <strong>condition</strong> so the postgres health check was actually getting monitored by the dependent service, e.g.:</p>
<pre><code class="lang-auto">    depends_on:
      db:
        condition: service_healthy
</code></pre>
<p>See SO post here: <a href="https://stackoverflow.com/questions/68941526/psql-not-found-in-script-for-trying-to-control-startup-and-shutdown-order-in-doc" class="inline-onebox" rel="noopener nofollow ugc">django - psql not found in script for trying to control startup and shutdown order in Docker Compose - Stack Overflow</a></p>
<p>Note that this isn’t officially supported for docker compose v3, but it seems to be unofficially supported. My editor gives me a warning on the <strong>condition</strong> statement. It works fine when running in Github, so I’m content for now.</p>

**Comment by fchabouis** [comment]
Best explanation I found about this problem is in this [blog post](https://laurent-bel.medium.com/waiting-for-postgresql-to-start-in-docker-compose-c72271b3c74a).

In short, for the web service, add a service health condition and a restart "always".

```
    depends_on:
      db:
        condition: service_healthy
    container_name: web
    restart: always
```

for the db, specify what "healthy" means:

```
  db:
    ports:
       - "5432:5432"
    ...
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: 5s
      timeout: 5s
      retries: 5
```
